File: get_forge_or_fabric_version.py
import os
import pathlib
import shutil
import json

def get_forge_or_fabric_version_from_manifest(path):
    with open(path, encoding="UTF-8") as f:
        data = json.load(f)
        modloaders = data["minecraft"]["modLoaders"]
        minecraft_version = data["minecraft"]["version"]
        
        for modloader in modloaders:
            if modloader.get("primary") == True:
                loader_id = modloader["id"].lower()

                # 1. Fabric Check
                if "fabric" in loader_id:
                    return "fabric", minecraft_version

                # 2. NeoForge Check (Wichtig: Vor Forge prüfen, da "forge" in "neoforge" enthalten ist)
                if "neoforge" in loader_id:
                    # Schneidet "neoforge-" ab (9 Zeichen)
                    return "neoforge", minecraft_version + "-" + modloader["id"][9:]

                # 3. Forge Check
                if "forge" in loader_id:
                    # Schneidet "forge-" ab (6 Zeichen)
                    return "forge", minecraft_version + "-" + modloader["id"][6:]

    return None, None

File: src/getForgeOrFabricVersion.ts
import fs from 'fs-extra';

export type ModLoaderKind = 'fabric' | 'forge' | 'neoforge';

export interface LoaderInfo {
    kind: ModLoaderKind;
    mcVersion: string;
    loaderVersion: string;
}

interface CurseModLoader {
    id?: string;
    primary?: boolean;
}

interface CurseManifest {
    minecraft?: {
        version?: string;
        modLoaders?: CurseModLoader[];
    };
}

/**
 * Parse Curseforge / serverpack-creator manifest.json and return loader info.
 *
 * Curse manifest mod loader ids look like:
 *   - "forge-47.2.0"
 *   - "neoforge-20.4.190"
 *   - "fabric-0.15.3"
 */
export async function getLoaderFromManifest(
    manifestPath: string
): Promise<LoaderInfo | null> {
    const raw = await fs.readFile(manifestPath, 'utf8');
    const data = JSON.parse(raw) as CurseManifest;

    const modloaders = data?.minecraft?.modLoaders || [];
    const minecraftVersion = data?.minecraft?.version;
    if (!minecraftVersion) return null;

    for (const ml of modloaders) {
        if (!ml.primary) continue;
        const id = ml.id || '';
        const lower = id.toLowerCase();

        // Order matters: check neoforge before forge so the "forge"
        // substring in "neoforge" doesn't misclassify.
        if (lower.startsWith('neoforge-')) {
            return {
                kind: 'neoforge',
                mcVersion: minecraftVersion,
                loaderVersion: id.slice('neoforge-'.length)
            };
        }
        if (lower.startsWith('forge-')) {
            return {
                kind: 'forge',
                mcVersion: minecraftVersion,
                loaderVersion: id.slice('forge-'.length)
            };
        }
        if (lower.startsWith('fabric-')) {
            return {
                kind: 'fabric',
                mcVersion: minecraftVersion,
                loaderVersion: id.slice('fabric-'.length)
            };
        }
    }
    return null;
}
